logstash-core/src/main/java/org/logstash/util/PasswordValidator.java

```diff
@@ -0,0 +1,128 @@
+package org.logstash.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+
+import java.util.List;
+
+/**
+ * A class to validate the given password string and give a reasoning for validation failures.
+ * Validation policies are based on complex password generation recommendation from several institutions
+ * such as NIST (https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-63b.pdf),
+ * OWASP (https://github.com/OWASP/www-community/blob/master/pages/OWASP_Validation_Regex_Repository.md), etc...
+ */
+public class PasswordValidator {
+
+    /**
+    A regex for special character inclusion.
+     */
+    private static final String SYMBOL_REGEX = ".*[~!@#$%^&*()_+|<>?:{}].*";
+
+    /**
+     A regex for upper case character inclusion.
+     */
+    private static final String UPPER_CASE_REGEX = ".*[A-Z].*";
+
+    /**
+     A regex for lower case character inclusion.
+     */
+    private static final String LOWER_CASE_REGEX = ".*[a-z].*";
+
+    /**
+     A regex for digit number inclusion.
+     */
+    private static final String DIGIT_REGEX = ".*\\d.*";
+
+    /**
+     A policy failure reasoning for empty password.
+     */
+    private static final String EMPTY_PASSWORD_REASONING = "Password must not be an empty.";
+
+    /**
+     A policy failure reasoning for password length.
+     */
+    private static final String LENGTH_REASONING = "Password must be length of between 8 and 20.";
+
+    /**
+     A policy failure reasoning if password does not contain special character(s).
+     */
+    private static final String SYMBOL_REASONING = "Password must contain at least one special character.";
+
+    /**
+     A policy failure reasoning if password does not contain upper case character(s).
+     */
+    private static final String UPPER_CASE_REASONING = "Password must contain at least one upper case.";
+
+    /**
+     A policy failure reasoning if password does not contain lower case character(s).
+     */
+    private static final String LOWER_CASE_REASONING = "Password must contain at least one lower case.";
+
+    /**
+     A policy failure reasoning if password does not contain digit number(s).
+     */
+    private static final String DIGIT_REASONING = "Password must contain at least one digit between 0 and 9.";
+
+    /**
+     * Validates given string against strong password policy and returns the list of failure reasoning.
+     * Empty return list means password policy requirements meet.
+     * @param password
+     * @return List of failure reasoning.
+     */
+    public List<String> validate(String password) {
+        List<String> reasons = Lists.newArrayList();
+        if (Strings.isNullOrEmpty(password)) {
```

Comment:
I see the solution has a set of predicate methods like `containsDigit` and `isValidLength` and then a set of conditional instructions to add the right reasoning.
I'd suggest we capture the validation + reasoning in an inner Validator class. Its `validate` method would either return nothing or the reason, which could then be folded/accumulated into a full sentence.
Havin a single Object that is able to perform the validation as output the reason makes it easier to add/remove/change reasons in the future.

Reply:
Thanks @jsvd ! Yes, this makes source **open to extend**.